lib/enmeshed/relationship_template.rb

```diff
@@ -0,0 +1,152 @@
+
+    def url
+      "nmshd://tr##{@truncated_reference}"
+    end
+
+    def app_store_link
+      Settings.omniauth.nbp.enmeshed.app_store_link
+    end
+
+    def play_store_link
+      Settings.omniauth.nbp.enmeshed.play_store_link
+    end
+
+    def qr_code
+      RQRCode::QRCode.new(url).as_png(border_modules: 0)
+    end
+
+    def qr_code_path
+      Rails.application.routes.url_helpers.nbp_wallet_qr_code_users_path(truncated_reference:)
+    end
+
+    def truncated_reference
+      raise ConnectorError.new('RelationshipTemplate has not been persisted yet') unless @truncated_reference
+
+      @truncated_reference
+    end
+
+    def remaining_validity
+      expires_at - Time.zone.now
+    end
+
+    def self.create!(nbp_uid:)
+      new(nbp_uid:).create!
+    end
+
+    def self.display_name_attribute
+      @display_name_attribute ||= Attribute::Identity.new(type: 'DisplayName', value: DISPLAY_NAME)
+    end
+
+    def self.allow_certificate_request
+      # i18n-tasks-use t('users.nbp_wallet.enmeshed.AllowCertificateRequest')
+      @allow_certificate_request ||= Attribute::Relationship.new(type: 'ProprietaryBoolean', key: 'AllowCertificateRequest', value: true)
+    end
+
+    def to_json(*)
+      {
+        maxNumberOfAllocations: 1,
+        expiresAt: expires_at,
+        content: {
+          metadata: {nbp_uid:},
+          '@type': 'RelationshipTemplateContent',
+          onNewRelationship: {
+            '@type': 'Request',
+            items: shared_attributes + relationship_attributes + required_attribute_queries,
+          },
+        },
+      }.to_json(*)
+    end
+
+    private
+
+    def shared_attributes
+      display_name_attribute = self.class.display_name_attribute
+      [
+        {
+          '@type': 'ShareAttributeRequestItem',
+          mustBeAccepted: true,
+          attribute: display_name_attribute.to_h,
+          sourceAttributeId: display_name_attribute.id,
+        },
+      ]
+    end
+
+    def relationship_attributes
+      if Settings.omniauth&.nbp&.enmeshed&.allow_certificate_request
+        [
+          {
+            '@type': 'CreateAttributeRequestItem',
+            mustBeAccepted: true,
+            attribute: self.class.allow_certificate_request.to_h,
+          },
+        ]
+      else
+        []
+      end
+    end
+
+    def required_attribute_queries
+      REQUIRED_ATTRIBUTES.map do |attr|
+        {
+          '@type': 'ReadAttributeRequestItem',
+          mustBeAccepted: true,
+          query: {
+            '@type': 'IdentityAttributeQuery',
+            valueType: attr,
+          },
+        }
+      end
+    end
+
+    def fetch_existing
+      template = Connector.fetch_existing_relationship_template truncated_reference
+      return unless template
+
+      populate_from_existing(template)
+    end
+
+    def populate_from_existing(template)
+      @nbp_uid = template.dig(:content, :metadata, :nbp_uid)
+      @expires_at = DateTime.parse(template[:expiresAt])
```

Comment:
I just want to point to typical issues with parsing values and invalid content to be parsed. But since this is all part of a well-defined API, we probably don't expect invalid values here and therefore do not need to handle the error, correct?

Reply:
@Mathis-Z and I discussed how to proceed here. We will add a JSON Schema validation, ensuring that we always have the required fields.

<hr>

I was doing some research on how this could look like. Luckily, the `json_schemer` gem also supports OpenAPI, so that we can simply write:

```rb
API_SCHEMA = JSONSchemer.schema(YAML.safe_load_file(Rails.root.join('lib/enmeshed/api.schema.yml'), permitted_classes: [Time]))
response = JSON.parse(File.read(Rails.root.join('spec/fixtures/files/enmeshed/valid_relationship_created.json')))
API_SCHEMA.valid?(response)
```

Based on the issues identified with the JSON file, I would use the YML file downloaded from
- https://doc-demo.is.enmeshed.eu/docs/yaml
- https://github.com/nmshd/connector/blob/main/src/modules/coreHttpApi/openapi.yml (has some nice comments and better formatting)

Reply:
I added the schema checking :+1:

Reply:
Thank you! This looks pretty good already, and I am happy you decided to validate a response with the expected object (`Connector::API_SCHEMA.schema('RelationshipTemplate')`). 👏 

Based on this implementation, I went ahead and ensured we are validating even more responses than before. In fact, with the new changes pushed, every response is first validated before being processed further.